src/Api/Upload.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace Kirby\Api;
+
+use Kirby\Cms\App;
+use Kirby\Filesystem\Dir;
+use Kirby\Filesystem\F;
+use Kirby\Toolkit\Str;
+
+/**
+ * The Upload class supports file uploads in the
+ * context of the API
+ *
+ * @package   Kirby Api
+ * @author    Nico Hoffmann <nico@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ * @since     4.3.0
+ */
+class Upload
+{
+	/**
+	 * Handle chunked uploads by merging all chunks
+	 * in the tmp directory and only returning the new
+	 * $source path to the tmp file once complete
+	 */
+	public static function chunk(
+		Api $api,
+		string $source,
+		string $filename
+	): string|null {
+		// if the file is uploaded in chunks
+		if ($isChunk = $api->requestHeaders('X-Upload-Chunk')) {
+			// ensure the directory exists
+			Dir::make($dir = static::tmp());
+
+			// append chunk to tmp file with unique filename (id suffix)
+			$id     = $api->requestHeaders('X-Unique-Upload-Id');
+			$chunk  = fopen($source, 'r');
+			$source = $dir . '/' . $filename;
```

Comment:
- Filename also needs to be filtered against directory traversal, e.g. with `basename($filename)`.
- Can we avoid overwriting the `$source` variable? IMO that makes the code quite confusing.

Reply:
We have to overwrite it at least at the end of the if block since we need to return `$source` either unaltered if it wasn't a chunk or altered after the last chunk.

Reply:
The end of the if block could IMO have its own return statement, it doesn't need to fall through.